tests/test_service.py

```diff
@@ -0,0 +1,127 @@
+
+import pytest
+import logging
+
+from plumbum import ProcessExecutionError, local
+from plumbum.cmd import docker
+from plumbum.machines.local import LocalCommand
+
+logger = logging.getLogger()
+
+CONTAINER_NAME = "docksockprox_test"
+SOCKET_PROXY = "127.0.0.1:2375"
+
+
+def _start_proxy(
+    container_name=CONTAINER_NAME,
+    socket_proxy=SOCKET_PROXY,
+    extra_args=None
+):
+    logger.info(f"Starting {container_name} with args: {extra_args}...")
+    docker(
+        "run",
+        "-d",
+        "--name", container_name,
+        "--privileged",
+        "-v", "/var/run/docker.sock:/var/run/docker.sock",
+        "-p", f"{socket_proxy}:2375",
+        extra_args,
+        "tecnativa/docker-socket-proxy",
+    )
+
+
+def _stop_and_delete_proxy(
+    container_name=CONTAINER_NAME,
+    socket_proxy=SOCKET_PROXY,
+):
+    logger.info(f"Removing {container_name}...")
+    docker(
+        "rm",
+        "-f",
+        container_name,
+    )
+
+
+def _query_docker_with_proxy(socket_proxy=SOCKET_PROXY, extra_args=None):
+    try:
+        _ret_code, stdout, stderr = docker.run(
+            (
+                "--host",
+                socket_proxy,
+                extra_args,
+            )
+        )
+    except ProcessExecutionError as result:
+        stdout = result.stdout
+        stderr = result.stderr
+    return stdout + stderr
```

Comment:
All this method is weird and unnecessary.

Weird because it slurps the exception and return code, so the caller cannot know if the process succeeded or failed.

Unnecessary because you can declare a partial local command instead:

```suggestion
_query_docker_with_proxy = docker["--host", SOCKET_PROXY]
```

Then, later:

```py
_query_docker_with_proxy(*extra_args)
```

---

Finally, given what I told you above about container names and ports, you'd better make this a fixture, so tests can use it and obtain a way to call the proxy. Pseudocode example:

```py
from plumbum.cmd import docker

@pytest.fixture
def docker():
    container, port = _start_proxy()
    try:
        with local.env(DOCKER_HOST=f"{container}:{port}"):
            yield docker
    finally:
        docker("container", "rm", "-f", container)
```

Reply:
Ok, I understand this improvements. 
However, they might bring some other problems and, at the same time, not introduce as many benefits.

> Weird because it slurps the exception and return code, so the caller cannot know if the process succeeded or failed.

The reason it captures the exception and return code is because the docker command will exit with a failure exit code if the operation has been forbidden, and 0 otherwise. Then, according to the command result, Plumbum will either return the `stdout` and `stderr` through the return values of the `run()` function, or through the result variable, placed on the exception thrown. The purpose of this function it to capture that **expected** exception thrown by the command.

> Unnecessary because you can declare a partial local command instead

Given what I said above, I don't really see the advantage of using this method instead of simply calling the `docker()` command. We will only be using this once in the tests anyway.

WDYT?

Reply:
> The purpose of this function it to capture that **expected** exception thrown by the command.

In such case, the test should use [pytest.raises](https://docs.pytest.org/en/reorganize-docs/new-docs/user/pytest_raises.html).



> Given what I said above, I don't really see the advantage of using this method instead of simply calling the `docker()` command. We will only be using this once in the tests anyway.

Well, I think that this system will be used a lot as tests grow, basically because most of what we have to test is making sure the container proxies the calls properly, so we need the proxy running.

However one thing is true: the proxy should be configurable on start, to be able to test it properly. So, maybe this would fit better as a helper context manager method. I imagine some test code like:

```python
import pytest
from plumbum import ProcessExecutionError
from plumbum.cmd import docker
from .conftest import start_proxy

def test_post_call():
    with start_proxy(POST=1):
        docker("container", "run", "--rm", "hello-world")
    with start_proxy(POST=0), pytest.raises(ProcessExecutionError):
        docker("container", "run", "--rm", "hello-world")
```